File: 2018/day3.py
#!python

import re

with open('d3.txt', 'r') as f:
    data = [ l.strip() for l in f.readlines() ]
    matrix = [ [], [] ]
    total = 0

    data = [ '#1 @ 82,901: 26x12' ]

    for d in data:
        m = re.match(r'#(\d+) \@ (\d+),(\d+): (\d+)x(\d+)', d)
        (claim_id, x_start, y_start, width, height) = map(int, m.groups())

        for x in range(x_start, x_start + width):
            for y in range(y_start, y_start + height):
                there = matrix[y][x]
                if there == '1':
                    total += 1
                matrix[y][x] = 'X' if there else '1'
    print(total)

File: 2018/day3.pl
#!perl
use warnings;
use v5.24;

my @data = (
  '#1 @ 1,3: 4x4',
  '#2 @ 3,1: 4x4',
  '#3 @ 5,5: 2x2',
);

open my $fh, '<', 'd3.txt' or die "couldn't read file: $!";
@data = map {; chomp; $_ } <$fh>;

my $matrix = [ [], [] ];
my $total = 0;

for my $d (@data) {
  my ($id, $x_start, $y_start, $width, $height) = $d =~ /#(\d+) \@ (\d+),(\d+): (\d+)x(\d+)/;

  for my $x ($x_start .. $x_start + $width - 1) {
    for my $y ($y_start .. $y_start + $height - 1) {
      my $there = $matrix->[$y][$x];
      $total++ if defined $there && $there eq '1';
      $matrix->[$y][$x] = $there ? 'X' : 1;
    }
  }
}

say "total: $total";

# Hard part

ID: for my $d (@data) {
  my ($id, $x_start, $y_start, $width, $height) = $d =~ /#(\d+) \@ (\d+),(\d+): (\d+)x(\d+)/;

  for my $x ($x_start .. $x_start + $width - 1) {
    for my $y ($y_start .. $y_start + $height - 1) {
      next ID if $matrix->[$y][$x] eq 'X';
    }
  }

  say "found one: $id";
  last ID;
}
